### PR TITLE
[ie/tenplay] Fix extractor

### LINKED ISSUE #10926: 10play - Website code changed breaks support
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Australia

### Provide a description that is worded well enough to be understood

Appears to be a site code change which has effected yt-dlp's ability to download from the site. Tested links that work yesterday which now no longer works.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
yt-dlp.exe -F https://10play.com.au/neighbours/episodes/season-42/episode-9107/tpv240902nzqyp -u PRIVATE -p PRIVATE -vU
[debug] Command-line config: ['-F', 'https://10play.com.au/neighbours/episodes/season-42/episode-9107/tpv240902nzqyp', '-u', 'PRIVATE', '-p', 'PRIVATE', '-vU']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version nightly@2024.09.02.232855 from yt-dlp/yt-dlp-nightly-builds [e8e6a982a] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.22631-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg N-116754-ga754ee0844-20240821 (setts), ffprobe N-116754-ga754ee0844-20240821
[debug] Optional libraries: Cryptodome-3.20.0, brotli-1.1.0, certifi-2024.08.30, curl_cffi-0.5.10, mutagen-1.47.0, requests-2.32.3, sqlite3-3.35.5, urllib3-2.2.2, websockets-13.0.1
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests, websockets, curl_cffi
[debug] Loaded 1832 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp-nightly-builds/releases/latest
Latest version: nightly@2024.09.02.232855 from yt-dlp/yt-dlp-nightly-builds
yt-dlp is up to date (nightly@2024.09.02.232855 from yt-dlp/yt-dlp-nightly-builds)
[TenPlay] Extracting URL: https://10play.com.au/neighbours/episodes/season-42/episode-9107/tpv240902nzqyp
[TenPlay] tpv240902nzqyp: Downloading JSON metadata
[TenPlay] tpv240902nzqyp: Getting bearer token
[TenPlay] tpv240902nzqyp: Downloading video JSON
[TenPlay] tpv240902nzqyp: Downloading webpage
ERROR: Invalid URL 'https://': No host supplied
Traceback (most recent call last):
  File "yt_dlp\networking\_requests.py", line 328, in _send
  File "requests\sessions.py", line 575, in request
  File "requests\sessions.py", line 484, in prepare_request
  File "requests\models.py", line 367, in prepare
  File "requests\models.py", line 444, in prepare_url
requests.exceptions.InvalidURL: Invalid URL 'https://': No host supplied

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "yt_dlp\YoutubeDL.py", line 1626, in wrapper
  File "yt_dlp\YoutubeDL.py", line 1761, in __extract_info
  File "yt_dlp\extractor\common.py", line 740, in extract
  File "yt_dlp\extractor\tenplay.py", line 96, in _real_extract
  File "yt_dlp\extractor\common.py", line 896, in _request_webpage
  File "yt_dlp\YoutubeDL.py", line 4165, in urlopen
  File "yt_dlp\networking\common.py", line 117, in send
  File "yt_dlp\networking\_helper.py", line 208, in wrapper
  File "yt_dlp\networking\common.py", line 340, in send
  File "yt_dlp\networking\_requests.py", line 360, in _send
yt_dlp.networking.exceptions.RequestError: Invalid URL 'https://': No host supplied
```

### PR DESCRIPTION
**Pull request by aarubui**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Switching to `vod.ten.com.au` API

Fixes #10926

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/tenplay.py
Comment (seproDev) [inline yt_dlp/extractor/tenplay.py:73]: The altId is required anyway, so let's just do:
```suggestion
        video = self._download_json(
            f'https://pubads.g.doubleclick.net/ondemand/hls/content/2489270/vid/{data["altId"]}/streams',
            content_id, data=b'')
```

File: yt_dlp/extractor/tenplay.py
Comment (seproDev) [inline yt_dlp/extractor/tenplay.py:76]: How about:
```suggestion
        formats, subtitles = self._extract_m3u8_formats_and_subtitles(video['stream_manifest'], content_id, 'mp4')
        for sub in video.get('subtitles', []):
            if ttml_url := sub.get('ttml'):
                subtitles.setdefault(sub.get('language') or 'en', []).append({
                    'url': ttml_url,
                    'ext': 'ttml',
                })
            if webvtt_url := sub.get('webvtt'):
                subtitles.setdefault(sub.get('language') or 'en', []).append({
                    'url': webvtt_url,
                    'ext': 'vtt',
                })
```

File: yt_dlp/extractor/tenplay.py
Comment (seproDev) [inline yt_dlp/extractor/tenplay.py:80]:
```suggestion
            'subtitles': subtitles,
```

File: yt_dlp/extractor/tenplay.py
Comment (seproDev) [inline yt_dlp/extractor/tenplay.py:69]: I assume you removed this due to login being no longer required. Is this the case for all content?

**Review by seproDev — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/tenplay.py
Comment (aarubui) [reply yt_dlp/extractor/tenplay.py:69]: Yes.

**Review by seproDev — APPROVED** [review]
Not ideal that we need to use the dynamic ad insertion URLs by Google, but I wasn't able to find any other way.
For downloading without ads, `--hls-split-discontinuity` and some manual effort is necessary, though still better than not being able to download at all.

**Comment by Ronnnny** [comment]
Is there anything that can help from https://github.com/matthuisman/slyguy.addons/tree/master/slyguy.10play
or https://forum.videohelp.com/threads/415165-[Release]-Ozivine-Australian-NZ-FTA-Downloader/page4#post2749536

File: yt_dlp/extractor/tenplay.py
Comment (seproDev) [inline yt_dlp/extractor/tenplay.py:85]: Imo. we can also add the resolution upgrade code after the geo restriction check:
```py
        # Attempt to get a higher quality stream
        m3u8_url = m3u8_url.replace(',150,75,55,0000', ',300,150,75,55,0000')
```

File: yt_dlp/extractor/tenplay.py
Comment (seproDev) [inline yt_dlp/extractor/tenplay.py:84]: I'd write this like so:
```suggestion
        video_data = self._download_json(
            f'https://vod.ten.com.au/api/videos/bcquery?command=find_videos_by_id&video_id={data["altId"]}',
            content_id, 'Downloading video JSON')
        m3u8_url = self._request_webpage(
            HEADRequest(traverse_obj(video_data, ('items', 0, 'HLSURL', {url_or_none}))),
            content_id, 'Checking stream URL').url
```

**Review by seproDev — COMMENTED** [review]
Please don't force push in the future. It makes reviewing more difficult.

**Comment by aarubui** [comment]
@seproDev It's hard to gauge your level of interest in this fix given you didn't check for yourself that logging in was no longer necessary, you didn't just commit your suggestions and you mentioned ad insertion which I wasn't experiencing but then you went to the trouble of looking up the `HLSURL` format. If I thought you cared enough to be reviewing at the time, naturally, I wouldn't have forced pushed.

So does that mean you actually got ads with the `doubleclick.net` videos? For which video? Or was it for all videos?

**Comment by seproDev** [comment]
I did a fair bit of checking for this PR, actually. Just like you, I want this PR (as well as any other PR) merged. However, not knowing the website, I wanted to make sure that removing the authentication code does not remove functionality that is still required.
Yes, I did see that for the couple videos I tried, no authentication was necessary. Are there paid videos I don't know about? Is there legacy content that still works differently? As maintainers, we rely on people that are familiar with the site for these kinds of questions.

> you didn't just commit your suggestions

I don't always just apply my suggestions. They are after all suggestions. If the other person is responsive, I usually leave it up to them to apply them. If they think the suggestions are good, it's a single button press in the web interface.

> you mentioned ad insertion

Yeah, that is the point of the doubleclick domain, which effectively proxies the manifest and inserts ads. See https://support.google.com/admanager/answer/7299051?hl=en
And yes, when downloading https://10play.com.au/neighbours-classics/episodes/season-2012/neighbours-9-jan-2012/tpv231004pgxmn I did get a video file that contained ads using the older version of this PR that used the doubleclick domain. There was one ad at the start and three ads throughout. Here the variant m3u8: [variant.m3u8.txt](https://github.com/user-attachments/files/16919959/variant.m3u8.txt)

**Comment by aarubui** [comment]
There are no paid videos. `memberGated` being a simple boolean affirms there's only a distinction between logged in and logged out.

Unless it's another maintainer, who's going to oppose suggestions? It's hard enough getting a PR looked at let alone getting it merged without fighting suggestions.

Thanks for the m3u8 attachment. Google DAI's not injecting anything for me so I thought 10play had the ads turned off. I guess we're depending on the API to keep the direct `fastly.net` playlists exposed.

**Comment by seproDev** [comment]
> Unless it's another maintainer, who's going to oppose suggestions? It's hard enough getting a PR looked at let alone getting it merged without fighting suggestions.

Yes, sometimes other maintainers might disagree, but it is also totally valid for the author of the PR to disagree. Like everyone, I do make mistakes. If you disagree or have concerns with a suggestion, you should voice that.

Regarding not getting your PRs looked at, we know that the backlog of PRs has been growing and that this is far from ideal.
There is, however, only so much time to go around. yt-dlp is a hobby project for all of us. Most of us have full-time jobs that actually put food on the table, so there is only so much time we can devote to yt-dlp.
I know that it is frustrating to not get a single reply for weeks on a PR that you put a lot of effort in to. We truly do want every PR reviewed and merged, but with currently only two people reviewing extractor PRs, it might take some time.

**Comment by billybanana80** [comment]
I pushed my [Ozivine]https://github.com/billybanana80/ozivine/tree/Ozivine_v1.5(url) solution out using the android endpoint as suggested by stabbedbybrick. Nice that you've picked up on his comments to fix this. Great job.

Will keep the web option up my sleeve though in case they decide to insert ads on the android side of things.

**Comment by aarubui** [comment]
> Yes, sometimes other maintainers might disagree, but it is also totally valid for the author of the PR to disagree. Like everyone, I do make mistakes. If you disagree or have concerns with a suggestion, you should voice that.

That makes me sound like I have a disagreement. I don't. Which makes you allowing room for feedback actually slow us down. It's not just "a single button press". I have to review the suggestions, run the tests and run `hatch fmt`. I can voice concerns even with maintainer commits.

> I know that it is frustrating to not get a single reply for weeks on a PR that you put a lot of effort in to. We truly do want every PR reviewed and merged, but with currently only two people reviewing extractor PRs, it might take some time.

I'm not frustrated at all. I already have the new code in my `yt_dlp_plugins` folder. So whatever you suggest, I would accept without exception. Anything to get this PR over with and stop it being my responsibility.

File: yt_dlp/extractor/tenplay.py
Comment (dirkf) [inline yt_dlp/extractor/tenplay.py:92]: Below info returned was not changed in this PR but could have some more type conditioning, like this:
```py
        return {
            'id': content_id,
            'formats': formats,
            'uploader': 'Channel 10',
            'uploader_id': '2199827728001',
            **traverse_obj(data, {
                'subtitles': ('captionUrl', {lambda x: None if x is None
                                          else {'en': [{'url': x}]}}),
                'id': ('altId', {str}),
                'duration': ('duration', {float_or_none}),
                'title': ('subtitle', {str}),
                'alt_title': ('title', {str}),
                'description': ('description', {str}),
                'age_limit': ('classification', {self._AUS_AGES.get}),
                'series': ('tvShow', {str}),
                'season_number': ('season', {int_or_none}),
                'episode_number': ('episode', {int_or_none}),
                'timestamp': ('published', {int_or_none}),
                'thumbnail': ('imageUrl', {url_or_none}),
            }),
        }
```

File: yt_dlp/extractor/tenplay.py
Comment (bashonly) [inline yt_dlp/extractor/tenplay.py:84]: no point in traversal here since the request is fatal
```suggestion
            HEADRequest(video_data['items'][0]['HLSURL']),
```

**Review by bashonly — APPROVED** [review]

File: yt_dlp/extractor/tenplay.py
Comment (bashonly) [inline yt_dlp/extractor/tenplay.py:64]: `x-forwarded-for` doesn't work with this site, so we want to add this back
```suggestion
    _GEO_BYPASS = False

```

**Review by bashonly — APPROVED** [review]

**Pull request merged**
